Demos/include/ApplicationBase.hpp

```diff
@@ -0,0 +1,598 @@
+// SPDX-FileCopyrightText: 2024 Vector Informatik GmbH
+//
+// SPDX-License-Identifier: MIT
+
+#include "silkit/SilKit.hpp"
+#include "silkit/services/logging/ILogger.hpp"
+#include "silkit/services/orchestration/all.hpp"
+#include "silkit/services/orchestration/string_utils.hpp"
+
+#include <string>
+#include <vector>
+#include <memory>
+#include <future>
+#include <thread>
+#include <unordered_set>
+#include <cctype>
+
+#include "SignalHandler.hpp"
+#include "CommandlineParser.hpp"
+
+using namespace SilKit;
+using namespace SilKit::Services::Orchestration;
+using namespace SilKit::Services::Can;
+using namespace SilKit::Services::Logging;
+using namespace SilKit::Util;
+
+using namespace std::chrono_literals;
+
+class ApplicationBase 
+{
+public:
+    ApplicationBase(std::string participantName)
+    {
+        _arguments.participantName = std::move(participantName);
+        _arguments.registryUri = "silkit://localhost:8500";
+        _arguments.runAsync = false;
+        _arguments.duration = 1000us;
+        _arguments.participantConfiguration = SilKit::Config::ParticipantConfigurationFromString("");
+
+        _commandLineParser = std::make_shared<CommandlineParser>();
+    }
+
+protected:
+    // Must be implemented by the actual application
+
+    // Add application specific command line arguments. Called during SetupCommandLineArgs.
+    virtual void AddCommandLineArgs() = 0;
+    // Evaluate application specific command line arguments. Called during SetupCommandLineArgs.
+    virtual void EvaluateCommandLineArgs() = 0;
+    
+    // Controller creates SIL Kit controllers / services here. 
+    virtual void CreateControllers() = 0;
+
+    // Boot up SIL Kit controllers here.
+    // Called in the CommunicationReadyHandler to ensure that messages emerging from the 
+    // controller creation reach other involved participants.
+    virtual void InitControllers() = 0;
+
+    // When running with the '--async' flag, this is called continuously in a separate worker thread.
+    // This thread is managed by the ApplicationBase. 
+    // The function will trigger as soon as the participant state is 'Running'. 
+    virtual void DoWorkAsync() = 0;
+
+    // When running with time synchronization (no '--async' flag), this is called in the SimulationStepHandler.
+    virtual void DoWorkSync(std::chrono::nanoseconds now) = 0;
+
+private:
+    // Command line parser
+    std::shared_ptr<CommandlineParser> _commandLineParser;
+
+    // Default command line argument identifiers (to allow exclusion)
+    enum DefaultArg
```

Comment:
Since you only use the enumerators as 'scoped' (e.g., `DefaultArg::Duration`) you should probably use `enum struct` or `enum class`. See [cppreference] for details.

If the enum is defined using just `enum`, each enumerator can be used 'unscoped' as well, e.g., `Duration == DefaultArg::Duration`.

```suggestion
    enum struct DefaultArg
```

[cppreference]: https://en.cppreference.com/w/cpp/language/enum